components/webcompat/content/browser/webcompat_exceptions_service.cc

```diff
@@ -190,8 +190,10 @@ void WebcompatExceptionsService::SetRules(
 }
 
 void WebcompatExceptionsService::SetRulesForTesting(
-  PatternsByWebcompatTypeMap patterns_by_webcompat_type) {
-    SetRules(std::move(patterns_by_webcompat_type));
+    const std::string& json_string) {
```

Comment:
this seems a bit more clunky than just directly defining `PatternsByWebcompatTypeMap` and passing it in? This service unfortunately uses the deprecated `LocalDataFilesObserver` which tightly couples the component updater to the service itself so there is no public method to test the parsing, but we're implicitly combining them here and ideally we should be unit testing the parsing separately so we can cover edge cases there.

Reply:
I think there's value in leaving this as a (nearly) end-to-end test? Where we start with JSON text and confirm that it results in the expected settings.

But I agree we could add more individual tests for parsing the JSON rule: https://github.com/brave/brave-browser/issues/45134